man/ncid-notify.1
.\" %W% %G%
.TH NCID-NOTIFY 1
.SH NAME
.B ncid-notify
- send a NCID notification to your Android or iOS device
.SH SYNOPSIS
ncid --no-gui --program
.B ncid-notify
.SH DESCRIPTION
The
.B ncid-notify
is a NCID notification program that uses NMA (Notify My Android),
Prowl, or both NMA and Prowl
It displays the Caller ID or message on a Android device, a iOS device,
or one of each.
.SH REQUIREMENTS
.TP
Android Device Requirements:
The "Notify My Android" app on your Android device.
.br
Registration at NMA website http://www.notifymyandroid.com/
.br
Generated API key to place in configuration file.
.TP
iOS Device Requirements:
The Prowl (Growl client for iOS) app from the app store
.br
Registration at Prowl website http://www.prowlapp.com/
.br
Generated API key to place in configuration file.
.SH FILES
/etc/ncid/conf.d/ncid-notify.conf
.br
/usr/share/ncid/ncid-notify
.br
/etc/ncid/key-prowl-api
.br
/etc/ncid/key-nma-api
.br
/etc/ncid/key-prowl-provider
.br
/etc/ncid/key-nma-developer
.SH SEE ALSO
ncidd.8,
ncid-modules.7,
ncid.1,
ncid-alert.1,
ncid-initmodem.1,
ncid-kpopup.1,
ncid-mythtv.1,
ncid-page.1,
ncid-samba.1,
ncid-skel.1,
ncid-speak.1,
ncid-wakeup.1,
ncid-yac.1


doc/CHANGES
ncid - Network CallerID

Version 0.87
    - README: updated
    - Makefile: added test: $(DOCDIR) does not exist, create it
    - All GPL license headers in files were updated to GPL 3
    - client/ncid.sh: logClock{} changed {^(CID|HUP|OUT)} to {^(?:CID|HUP|OUT)}
                      added "Line Labels" in help menu and labInfo {}
                      checkTypes() added new types and changed type numbers
                      getCID(): added new types and changed if statements
                                for new Type numbers
    - debian/Makefile: updated
    - debian/README.Debian: updated
    - debian/*.init: removed "S" from default STOP
    - debian/wc2ncid.init: removed configuration stuff, changed sbin to sbin
    - debian/yac2ncid.init: removed configuration stuff, changed sbin to sbin
    - debian/rn2ncid.init: new
    - doc/INSTALL-Ubuntu: updated
    - doc/INSTALL-Raspbian: updated
    - doc/LICENSE: updated to GPL 3
    - doc/Setup-rn2ncid: new
    - doc/Verbose-rn2ncid: new
    - Fedora/Makefile: updated
    - Fedora/README.Fedora: updated
    - Fedora/rn2ncid.init: new
    - Fedora/rn2ncid.service: new
    - FreeBSD/Makefile: updated
    - FreeBSD/README.FreeBSD: updated
    - FreeBSD/rn2ncid.rc: new
    - gateway/Makefile: added $(EXTRA_CFLAGS) to CFLAGS definition
    - gateway/README.Gateways: renamed to README.gateways and updated
    - gateway/*.[ch]: updated license in headers
    - gateway/*.pl: updated license in comments
    - gateway/sip2ncid.h: added UNAUTHORIZED define
    - gateway/sip2ncid.c: processPackets() added UNAUTHORIZED section
    - gateway/wc2ncid.conf: - improved comment wording
                            - changed nciddelay to delay
    - gateway/wc2ncid.pl: - moved test mode just after Command Line Processing
                          - moved logfile and config file messages to
                            after start message
                          - added cleanup(), called by errorExit() & sigHandle()
                          - added delay option to pod documentation wording,
                            improved some pod wording
                          - changed $nciddelay to $delay, added $Delay,
                            added --delay|-d option
    - gateway/rn2ncid.pl: new
    - gateway/rn2ncid.config: new
    - Mac/Makefile: updated
    - Mac/README.Mac: updated
    - Mac/net.sourceforge.rn2ncid.plist: new
    - man/Makefile: updated
    - man/ncid-*.1: updated
    - man/ncid.1: updated
    - man/ncid.conf.5: updated
    - man/ncidd.8: updated
    - man/yac2ncid.1: updated
    - man/rn2ncid.conf.5: new
    - modules/ncid-*.sh: updated
    - modules/ncid-*.conf: updated
    - modules/ncid-notify.sh: changed line one from /bin/sh to /bin/bash
                              minor changes to verbose output and comments
    - modules/ncid-speak.sh: added "AreaCodeLength" code to speak the area
                             code instead or "NO NAME" or entire number
    - modules/ncid-speak.conf-in: added "AreaCodeLength" and comments
    - server/ncidd.c: writeLog() - moved "logMsg(LEVEL3, msgbuf);" to start
                      formatCID() - moved cidnoname test inside
                                    if (!(cid.status & CIDNMBR))
                                 - added PID:
                      added BLKLINE, NOTLINE, PIDLINE to *lineTags[]
                      doPoll: - added code for a NOT: message
    - server/ncidd.h: added defines for PID, PIDLINE, NOTLINE, CALLPID
    - test/README.test: updated
    - test/*: renamed test programs and test data
    - test/test-rn2ncid: new
    - test/test-rn2ncid.data: new
    - TiVo/ncid-tivo.sh:  if no line label, display $CIDNAME/$CIDNMBR
                          else display "$CIDNAME $CIDNMBR/$CIDTYPE $CIDLINE"

Version 0.86.1
    - debian/changelog: updated
    - debian/control: for package ncid-client added "curl" to Depends

Version 0.86
    - README: updated
              added redhat: and redhat-install:
    - attic/*: new
    - client/ncid.sh: added more doVerbose statements and levels
                      changed 300* doVerbose message to any one of four
                      added line count to "End of call log" message
                      removed the --raw | -R option
                      fixed and improved call on a ring number
                      fonts used depend on platform and OS
                      allow resizing of window
                      allow font size changes
                      time format in history window can be either 12 and 24
                      new menu called Preferences for Fonts and time format
                      menu item to save window size
                      menu item to save window size and location
                      added the "-h | --help" option
                      added the "-v | --version" option
                      added the "-w | --wakeup" option
                      getCID{}: added calls to ncid-wakeup module
                      changed "-V" to be the "-v | --verbose" option
                      makeWindow{}: improved widget and menu code
                      added saveSize{} write_rc_file{} logFont{} changeFont{}
                      getField{}: changed regex to include "*" and "-"
    - client/ncid.conf-in: removed MsgFlag variable and comments
                           removed Raw variable and comments
                           added comments to CallOnRing section
                           added comments to Ring Options section
                           new "WakeUp" section
                           new "WrapLines" section
    - debian/Makefile: updated
    - debian/wc2ncid.init: new
    - debian/ncidsip.init: removed
    - doc/CONTRIBUTORS: updated
    - doc/NCID-SDK.odt: updated
    - doc/Setup-ncidsip: removed
    - doc/Setup-sip2ncid: updated
    - doc/Setup-wc2ncid: new
    - doc/INSTALL-Redhat: new
    - doc/INSTALL*: updated
    - doc/INSTALL-RPi: new
    - Fedora/Makefile: updated

modules/ncid-notify.sh
#!/bin/bash

# ncid-notify
# usage: ncid --no-gui --program ncid-notify

# Last modified: Wed May 29, 2013

# sends a NCID notification to a iOS device or a Android device

# Requirement for iOS:
# The Prowl (Growl client for iOS) app from the app store
# Free registration at the Prowl website <http://www.prowlapp.com/>
# Generated API key to place in configuration file.

# Requirement for Android
# The "Notify My Android" app on your Android device.
# Free registration at the NMA website <http://notifymyandroid.appspot.com/>
# Generated API key to place in configuration file.

# Module Requires:
# Generated API key from Prowl or NMA (Notify My Android)
# bash 2.0 or higher (hopefully TiVo has at least bash 2.0)
# curl
# CA certificates

# input is always 6 lines
#
# if input is from a call:
# input: DATE\nTIME\nNUMBER\nNAME\nLINE\nTYPE\n
#
# if input is from a message
# the message is in place of NAME:
# input: \n\n\n<MESSAGE>\n\nMSG\n

ConfigDir=/usr/local/etc/ncid/conf.d
ConfigFile=$ConfigDir/ncid-notify.conf

# default api key for each service
notify_apikey_prowl=
notify_apikey_nma=

# default api key file for each service
notify_apikeyfile_prowl=$ConfigDir/key-prowl-api
notify_apikeyfile_nma=$ConfigDir/key-nma-api

# default high volume key for each service (optional)
notify_hvolkey_prowl=""
notify_hvolkeyfile_prowl=$ConfigDir/key-prowl-provider
notify_hvolkey_nma=""
notify_hvolkeyfile_nma=$ConfigDir/key-nma-developer

# default protocol for each service
notify_protocol_prowl="https"
notify_protocol_nma="https"

# default priority level
notify_priority=0

# default time format
notify_clock="12"

# default call types to send notification
notify_types="CID"

# default 'application', 'event' and 'notification'
# by service

# Prowl defaults:
notify_application_prowl='ncid'
notify_event_prowl='$CIDDESC'
notify_notification_prowl='$CIDNAME $CIDNMBR $CIDLINE'

# NMA defaults: 
notify_application_nma='NCID \(${CIDTYPE}\)'
notify_event_nma='$CIDNAME $CIDNMBR $CIDLINE'
notify_notification_nma='$CIDDESC: $CIDDATE $CIDTIME'

# default URL
notify_url=

[ -f $ConfigFile ] && . $ConfigFile

# place scalar service-specific variables from ncid-notify.conf into
# an indexed array to make it easier to deal with multiple services
# service is either 'prowl' or 'nma' (Notify My Android)

     notify_service[1]="prowl"

      notify_apikey[1]="$notify_apikey_prowl"
  notify_apikeyfile[1]="$notify_apikeyfile_prowl"
     notify_hvolkey[1]="$notify_hvolkey_prowl"
 notify_hvolkeyfile[1]="$notify_hvolkeyfile_prowl"
    notify_protocol[1]="$notify_protocol_prowl"
 notify_application[1]="$notify_application_prowl"   
       notify_event[1]="$notify_event_prowl"
notify_notification[1]="$notify_notification_prowl"

	  
     notify_service[2]="nma"

      notify_apikey[2]="$notify_apikey_nma"
  notify_apikeyfile[2]="$notify_apikeyfile_nma"
     notify_hvolkey[2]="$notify_hvolkey_nma"
 notify_hvolkeyfile[2]="$notify_hvolkeyfile_nma"
    notify_protocol[2]="$notify_protocol_nma"
 notify_application[2]="$notify_application_nma"   
       notify_event[2]="$notify_event_nma"
notify_notification[2]="$notify_notification_nma"

# determine settings for each service
count=0
for i in "${notify_service[@]}"
do
    count=$(($count + 1))
    case $i in
            "") continue;; # handle sparse array element
         prowl) option[$count]="-F"
                www[$count]=${notify_protocol[$count]}://prowl.weks.net/publicapi/add;;
           nma) option[$count]="-d"
                www[$count]=${notify_protocol[$count]}://notifymyandroid.appspot.com/publicapi/notify;;
             *) exit 1;;
    esac
done

# if notify_apikeyfile exists & notify_apikey blank, get api key from file
count=0
found=""
for i in "${notify_service[@]}"
do
    count=$(($count + 1))
	[ -z "$i" ] && continue # handle sparse array element
    [ -f "${notify_apikeyfile[$count]}" -a ! "${notify_apikey[$count]}" ] &&
        notify_apikey[$count]=`cat ${notify_apikeyfile[$count]}`
    # api key is required
    if [ "${notify_apikey[$count]}" ] 
	   then
	   found=1
	else
	   notify_service[$count]="" # no api key so clear this service from being used
	fi
done

# must have at least one service
[ -z "$found" ] && exit 1
	
# similarly for the optional high volume keys	
count=0
for i in "${notify_service[@]}"
do
    count=$(($count + 1))
    [ -z "$i" ] && continue # handle sparse array element
    [ -f "${notify_hvolkeyfile[$count]}" -a ! "${notify_hvolkey[$count]}" ] &&
        notify_hvolkey[$count]=`cat ${notify_hvolkeyfile[$count]}`
done
	
read CIDDATE
read CIDTIME
read CIDNMBR
read CIDNAME
read CIDLINE
read CIDTYPE

# Only send notification if current call type is known
found=""
for i in $notify_types
do
    [ $i = "$CIDTYPE" ] && \
    {
        case $i in
            CID) CIDDESC="Incoming Call";;
            OUT) CIDDESC="Outgoing Call";;
            HUP) CIDDESC="Blacklisted Call Hangup";;
            BLK) CIDDESC="Blacklisted Call Blocked";;
            MSG) CIDDESC="Message";
		         CIDDATE=`date '+%m/%d/%Y'`;
		         CIDTIME=`date '+%H:%M'`;;
              *) CIDDESC='New Call Type: \(${CIDTYPE}\)';;
        esac
        found=1
        break
    }
done

# exit if call type is unknown
[ -z "$found" ] && exit 1

# determine if url should be used
[ "$notify_url" -a "$CIDTYPE" = "CID" ] && \
{
    # no safeguard for number being word(s)
    # set clickable url to the url + raw number
    rawnum=${CIDNMBR//-/}
    clickable_url=$notify_url$rawnum
}

[ "$CIDTIME" -a "$notify_clock" = "12" ] && \
{
    # convert time from 24 hour format to 12 hour format
    hour=${CIDTIME/:*/}
    minute=${CIDTIME/*:/}
    [ -z "$hour" ] && { hour=12 ampm=PM; }
    [ $hour -gt 11 ] && { hour=$(($hour - 12)); ampm=PM; } || ampm=AM
    notify_time="$hour:$minute $ampm"
} || notify_time=$CIDTIME

# reset CIDTIME to $notify_time so the 'eval' below will use the newly
# reformatted 12 hour time
CIDTIME=$notify_time

# send notification
count=0
for i in "${notify_service[@]}"
do
    count=$(($count + 1))
	[ -z "$i" ] && continue # handle sparse array element

	 notify_application[$count]=`eval echo ${notify_application[$count]}`
	       notify_event[$count]=`eval echo ${notify_event[$count]}`
	notify_notification[$count]=`eval echo ${notify_notification[$count]}`
	
	# providerkey  used by Prowl
	# developerkey used by NMA
	
    curl \
        ${option[$count]} apikey="${notify_apikey[$count]}" \
        ${option[$count]} application="${notify_application[$count]}" \
        ${option[$count]} event="${notify_event[$count]}" \
        ${option[$count]} description="${notify_notification[$count]}" \
		${option[$count]} priority="$notify_priority" \
    	${option[$count]} providerkey="${notify_hvolkey[$count]}" \
        ${option[$count]} developerkey="${notify_hvolkey[$count]}" \
        ${option[$count]} url="$clickable_url" \
        ${www[$count]}
done

exit 0
